File: src/904.java
class Solution {
    public int totalFruit(int[] tree) {
        Map<Integer, Integer> baskets = new HashMap<>();
        int start = 0, max = 0;

        for (int end = 0; end < tree.length; end++) {
            int endFruit = tree[end];
            baskets.put(endFruit, baskets.getOrDefault(endFruit, 0) + 1);

            while (start < end && baskets.size() > 2) {
                int startFruit = tree[start];
                baskets.put(startFruit, baskets.get(startFruit) - 1);
                if (baskets.get(startFruit) < 1) baskets.remove(startFruit);
                start++;
            }

            max = Math.max(max, end - start + 1);
        }

        return max;
    }
}

File: src/main/python/904.py
from typing import List

class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        n = len(tree)
        baskets = [0] * n
        maximum = end = start = cnt = size = 0

        while (end < n):
            endFruit = tree[end]
            if baskets[endFruit] == 0:
                size += 1
            baskets[endFruit] += 1
            cnt += 1
            end += 1

            while (size > 2):
                startFruit = tree[start]
                if baskets[startFruit] == 1:
                    size -= 1
                baskets[startFruit] -= 1
                cnt -= 1
                start += 1

            maximum = max(maximum, cnt)

        return maximum
